mbed_targets/_internal/target_attribute_hierarchy_parsers/overriding_attribute_parser.py

```diff
@@ -0,0 +1,95 @@
+"""Functions for parsing the inheritance for overriding attributes.
+
+Accumulating attributes are defined and can be overridden further down the hierarchy.
+The hierarchy is determined as 'depth-first' in multiple inheritance.
+eg. Look first for the attribute in the current target. If not found,
+look for the attribute in the first target's parent, then in the parent of the parent and so on.
+If not found, look for the property in the rest of the target's parents, relative to the
+current inheritance level.
+
+This means a target on a higher level could potentially override one on a lower level.
+"""
+from collections import deque
+from functools import reduce
+from typing import Dict, List, Any, Deque
+
+from mbed_targets._internal.target_attribute_hierarchy_parsers.accumulating_attribute_parser import (
+    ALL_ACCUMULATING_ATTRIBUTES,
+)
+
+
+def _targets_override_hierarchy(all_targets_data: Dict[str, Any], target_name: str) -> List[dict]:
+    """List all ancestors of a target in order of overriding inheritance (depth-first).
+
+    Using a depth-first traverse of the inheritance tree, return a list of targets in the
+    order of inheritance, starting with the target itself and finishing with its highest ancestor.
+
+    Args:
+        all_targets_data: a dictionary representation of all the data in a targets.json file
+        target_name: the name of the target we want to calculate the attributes for
+
+    Returns:
+        A list of dicts representing each target in the hierarchy.
+    """
+    override_order: List[dict] = []
+
+    still_to_visit: Deque[dict] = deque()
+    still_to_visit.appendleft(all_targets_data[target_name])
+
+    while still_to_visit:
```

Comment:
We should factor out the common code from here and `_targets_accumulate_hierarchy` into a more generic generator function.

Reply:
I thought that might make it even more impenetrable 😂

Reply:
I've had a think about this and I think we shouldn't try and factor out too much. This is a very self-contained algorithm, I don't want to have to be looking between files working out what is part of the shared code and what isn't. I also don't want the two parsers to be reliant on each other. I want to make it very easy in the future to change either/both of their functionality (I am not sure we want there to be two different ways of finding the hierarchy anyway).

Reply:
I still think we should definitely factor out the repeated code here. It will make it easier to understand if it's wrapped in a nicely named generator and reused in these two functions. In some _rare_ cases it's ok to ignore DRY and repeat some code. I don't think this is one of those situations though.

Reply:
After IRL walk through of the code I think that factoring out will add more complexity. Fingers crossed this can be simplified the correct way in the near future - by only having one way of inheriting. 🤞